compiler/rustc_target/src/spec/abi.rs

```diff
@@ -101,7 +97,40 @@ pub fn all_names() -> Vec<&'static str> {
 impl Abi {
     #[inline]
     pub fn index(self) -> usize {
-        self as usize
+        // N.B., this ordering MUST match the AbiDatas array above.
```

Comment:
Oof, this code is annoying =) I wonder if it's worth just having `C` and `Cunwind` instead. I guess that's less good in most other ways. Maybe add a `debug_assert` like

```rust
let i = match self { .. };
debug_assert!(AbiDates[self] == i);
i
```

Reply:
(Honestly, I wonder if we could do `AbiDatas.indexof(self)` or something like that...also `indexof` is like Java or something... but you get the idea.)

Reply:
> I wonder if it's worth just having C and Cunwind instead. I guess that's less good in most other ways. **[..]**

I thought about this as well! It would certainly make the changes specific to this file much easier :slightly_smiling_face: The primary reason I found that to be an unsatisfactory route forward was that RFC 2945 also includes [other `unwind` ABI strings](https://github.com/rust-lang/rfcs/blob/master/text/2945-c-unwind-abi.md#other-unwind-abi-strings).

`stdcall-unwind`, `system-unwind`, and `thiscall-unwind` are specified currently, but the RFC also states that "_More unwind variants of existing ABI strings may be introduced, with the same semantics, without an additional RFC._"

Adding more `FooUnwind` variants would certainly be _possible_, but it felt like it would make the process of introducing these new unwinding variants much more prone to error. It would require manual auditing of our `match` statements, and the number of variants could become tough to manage.

The `debug_assert!` is an excellent idea, I will add an additional commit with this logic. I certainly bumped my head against an error that this assertion would have prevented while working on this PR! :angel: